# HG changeset patch
# User Iain Murray  <iain@iainmurray.net>
# Date 1300395499 25200
#      Thu Mar 17 13:58:19 2011 -0700
# Node ID 77b14e634166d5d3951fca61b18213c82faf41aa
# Parent  85e87b865f71cdc2487679670aa77f5c0f26373f
Replace nprocs with nproc function.  Use gnulib module for portability across platforms.

diff --git a/ChangeLog b/ChangeLog
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,8 +1,12 @@
+2011-03-17  Iain Murray  <iain@iainmurray.net>
+
+	* bootstrap.conf (gnulib_modules): Include nproc in the list.
+
 2011-03-08  Rik  <octave@nomad.inbox5.com>
 
 	* mk-opts.pl: Recode using more modern Perl syntax.
 	Use my, not local, for lexically-scoped variables.
 	Use ALL_CAPITALS for global variable names.
 	Change code to remove any warnings from Perl lint (-w).
 	Inline small subroutines for better readability.
 	Use here documents and qq operator to avoid excessive backlashing of ".
diff --git a/bootstrap.conf b/bootstrap.conf
--- a/bootstrap.conf
+++ b/bootstrap.conf
@@ -31,16 +31,17 @@ gnulib_modules="
   glob
   link
   lstat
   mkdir
   mkfifo
   mkstemp
   mktime
   nanosleep
+  nproc
   pathmax
   progname
   readlink
   rename
   rmdir
   round
   roundf
   sigaction
diff --git a/doc/ChangeLog b/doc/ChangeLog
--- a/doc/ChangeLog
+++ b/doc/ChangeLog
@@ -1,8 +1,13 @@
+2011-03-17  Iain Murray  <iain@iainmurray.net>
+
+	* interpreter/system.txi: Remove nprocs and nprocs_conf functions
+	from documentation and replace with nproc.
+
 2010-03-16  Rik  <octave@nomad.inbox5.com>
 
 	* interpreter/system.txi: Add nproc and nproc_conf functions to
 	documentation.
 
 2010-03-03  Rik  <octave@nomad.inbox5.com>
 
 	* interpreter/matrix.txi: Deprecate is_duplicate_entry.
diff --git a/doc/interpreter/system.txi b/doc/interpreter/system.txi
--- a/doc/interpreter/system.txi
+++ b/doc/interpreter/system.txi
@@ -465,19 +465,17 @@ referred to as a @var{grp_struct}.
 
 @node System Information
 @section System Information
 
 @DOCSTRING(computer)
 
 @DOCSTRING(uname)
 
-@DOCSTRING(nprocs)
-
-@DOCSTRING(nprocs_conf)
+@DOCSTRING(nproc)
 
 @DOCSTRING(ispc)
 
 @DOCSTRING(isunix)
 
 @DOCSTRING(ismac)
 
 @DOCSTRING(isieee)
diff --git a/src/ChangeLog b/src/ChangeLog
--- a/src/ChangeLog
+++ b/src/ChangeLog
@@ -1,8 +1,15 @@
+2011-03-17  Iain Murray  <iain@iainmurray.net>
+
+	* DLD-FUNCTIONS/nprocs.cc: Delete file.
+	* DLD-FUNCTIONS/nproc.cc: New file.  New function nproc provided by
+	gnulib.
+	* DLD-FUNCTIONS/module-files: Add nproc.cc.
+
 2011-03-16  Iain Murray  <iain@iainmurray.net>
 
 	* DLD-FUNCTIONS/nprocs.cc: New file.
 	* DLD-FUNCTIONS/module-files: Add nprocs.cc.
 	Expose nprocs and nprocs_conf from gnulib.
 
 2010-03-15  Marco Caliari  <marco.caliari@univr.it>
 
diff --git a/src/DLD-FUNCTIONS/module-files b/src/DLD-FUNCTIONS/module-files
--- a/src/DLD-FUNCTIONS/module-files
+++ b/src/DLD-FUNCTIONS/module-files
@@ -50,17 +50,17 @@ kron.cc
 lookup.cc
 lsode.cc
 lu.cc
 luinc.cc
 matrix_type.cc
 max.cc
 md5sum.cc
 mgorth.cc
-nprocs.cc
+nproc.cc
 onCleanup.cc
 pinv.cc
 qr.cc
 quad.cc
 quadcc.cc
 qz.cc
 rand.cc
 rcond.cc
diff --git a/src/DLD-FUNCTIONS/nprocs.cc b/src/DLD-FUNCTIONS/nproc.cc
rename from src/DLD-FUNCTIONS/nprocs.cc
rename to src/DLD-FUNCTIONS/nproc.cc
--- a/src/DLD-FUNCTIONS/nprocs.cc
+++ b/src/DLD-FUNCTIONS/nproc.cc
@@ -1,13 +1,12 @@
 /*
 
 Copyright (C) 2011 Iain Murray
 
-
 This file is part of Octave.
 
 Octave is free software; you can redistribute it and/or modify it
 under the terms of the GNU General Public License as published by the
 Free Software Foundation; either version 3 of the License, or (at your
 option) any later version.
 
 Octave is distributed in the hope that it will be useful, but WITHOUT
@@ -21,56 +20,71 @@ along with Octave; see the file COPYING.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
 
 #include "defun-dld.h"
-#include "sys/sysinfo.h"
+#include "nproc.h"
 
-DEFUN_DLD (nprocs, args, nargout,
+DEFUN_DLD (nproc, args, nargout,
    "-*- texinfo -*-\n\
-@deftypefn {Loadable Function} {} nprocs ()\n\
-Return the number of available processors.\n\
-@seealso{nprocs_conf}\n\
+@deftypefn  {Loadable Function} {} nproc ()\n\
+@deftypefnx {Loadable Function} {} nproc (@var{query})\n\
+Return the current number of available processors.\n\
+\n\
+If called with the optional argument @var{query}, modify how processors\n\
+are counted as follows:\n\
+@table @code\n\
+@item all\n\
+total number of processors.\n\
+\n\
+@item current\n\
+processors available to the current process.\n\
+\n\
+@item overridable\n\
+likewise, but overridable through the OMP_NUM_THREADS environment variable.\n\
+@end table\n\
 @end deftypefn")
 {
   octave_value retval;
 
   int nargin = args.length ();
 
-  if (nargin != 0 || (nargout != 0 && nargout != 1))
+  if ((nargin != 0 && nargin != 1) || (nargout != 0 && nargout != 1))
     {
       print_usage ();
       return retval;
     }
 
-  retval = get_nprocs ();
+  nproc_query query = NPROC_CURRENT;
+  if (nargin == 1)
+    {
+      std::string arg = args(0).string_value ();
+
+      std::transform (arg.begin (), arg.end (), arg.begin (), tolower);
+
+      if (arg == "all")
+        query = NPROC_ALL;
+      else if (arg == "current")
+        query = NPROC_CURRENT;
+      else if (arg == "overridable")
+        query = NPROC_CURRENT_OVERRIDABLE;
+      else
+        {
+          error ("nproc: invalid value for QUERY");
+          return retval;
+        }
+    }
+
+  retval = num_processors (query);
 
   return retval;
 }
 
-DEFUN_DLD (nprocs_conf, args, nargout,
-   "-*- texinfo -*-\n\
-@deftypefn {Loadable Function} {} nprocs_conf ()\n\
-Return the number of number of processors the operating system has\n\
-configured.  This number may be less than the total available as reported by\n\
-@code{nprocs}.\n\
-@seealso{nprocs}\n\
-@end deftypefn")
-{
-  octave_value retval;
+/*
 
-  int nargin = args.length ();
+%% Must always report at least 1 cpu available
+%!assert (nproc () >= 1);
 
-  if (nargin != 0 || (nargout != 0 && nargout != 1))
-    {
-      print_usage ();
-      return retval;
-    }
-
-  retval = get_nprocs_conf ();
-
-  return retval;
-}
-
+*/
